src/Chapter3.hs

```diff
@@ -480,6 +512,31 @@ After defining the city, implement the following functions:
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
 
+data House = House Int deriving (Show) -- house with number of people in it
+data Church = Church deriving (Show)
+data Library = Library deriving (Show)
+data Castle = Castle String deriving (Show)
+data City = City {
+  castle :: Maybe Castle,
+  wall :: Bool,
+  building :: Either Church Library,
+  houses :: [House]
+} deriving (Show)
+
+buildCastle :: City -> String -> City
+buildCastle city name = city{castle = Just $ Castle name}
+
+buildHouse :: City -> Int -> City
+buildHouse c@City{houses=h} num = c{houses=House num : h}
+
+buildWalls :: City -> Either String City
+buildWalls c =
+  case castle c of
+    Nothing -> Left "No castle in the city!"
+    Just _ ->
+      case foldl (\acc (House n) -> acc+n) 0 (houses c) of
+        n | n < 10 -> Left "Not enough people in the city!"
+        _ -> Right c{wall=True}
```

Comment:
That's very good modelling of the result type with sum types. Great work :+1: 

If you want to improve the code even further, you can introduce a more precise custom data types:

```haskell
data BuildWallsResult
    = NoCastle
    | NotEnoughPeople
    | NewCity City
```

With a custom data type, it's more convenient to perform unit testing of various corner cases, rather than with comparing strings.